### ISSUE TITLE
Implement debugger

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#535]
The work has started: https://github.com/intellij-rust/intellij-rust/tree/fc2cfb8e0fc76984495a34ae6a83cbb37188c699/debugger

## State of the art

The debugger is officially available in CLion since 2018.1. Note, CLion doesn't support debugging with msvc toolchains, so if you are using **Windows**, you need: 
* install gnu rust toolchain (`rustup install stable-gnu`)
* setup gnu (Cygwin or MinGW) toolchains in CLion (`Preferences > Build, Execution, Deployment > Toolchains`).

To debug you just need to:
* Run debug command via gutter icon
![2018-05-28 11 46 26](https://user-images.githubusercontent.com/2539310/40606076-e42759be-626c-11e8-816a-cf002e9fb9db.gif)
* Or create `Cargo Command` run configurtion
<img width="817" alt="screen shot 2018-05-28 at 11 50 04" src="https://user-images.githubusercontent.com/2539310/40606199-508756a4-626d-11e8-8c9d-23c07d201704.png">
and run debugging via debug action

![2018-05-28 11 52 16](https://user-images.githubusercontent.com/2539310/40606324-b0ca35d6-626d-11e8-8ff2-4282ffa30c59.gif)

<details>
<summary>Old content</summary>

## Old content
If you use CLion and the **nightly Cargo**, you should be able to Debug Cargo configuration:

https://zippy.gfycat.com/AcidicMediocreCoati.webm

A major missing piece is that we don't pass command line arguments to the debuggee yet (this requires some work on the Cargo side of things first). 

## Old content 2

Currently for technical reasons ([commit message](https://github.com/intellij-rust/intellij-rust/commit/d464a1df6c6e9eb7efdc8e1e75d62818dc3ee5d9)) debugger is implemented as a separate plugin.

The current build can be download from TeamCity: [Download](https://teamcity.jetbrains.com/repository/download/IntellijIdeaPlugins_Rust_BuildDebugger/.lastSuccessful/intellij-rust-debugger-0.0.1.%7Bbuild.number%7D.zip).  It may work with **CLion** 2016.3.2 and intellij-rust 0.1.0.1611. 

Note that this is super unstable, debugger can invoke any kind of undefined behavior, including, but not limited to, summoning of nasal daemons. Most importantly, rust debugger may (though in theory it should not) break C++ debugging.

Nonetheless, step into and local variables view more or less work :) 

![debugger](https://cloud.githubusercontent.com/assets/1711539/22738874/59311ec0-ee1a-11e6-8203-f8ab98b93e31.png)

As you can see on the screenshot, at the moment you unfortunately have to create a separate `Cargo debug` run configuration type (as opposed to the usual `Run cargo command`). Also, you only can debug binaries now :(
</details>

**Comment by laktak** [comment]
Getting a good debugger for Rust would be a great help, especially as [the current options are rather limited](https://www.rust-lang.org/en-US/faq.html#debugging).

**Comment by mfarrugi** [comment]
How would the rust debugger be implemented? Is the CLion C/++ debugger more than a wrapper of GDB?

**Comment by matklad** [comment]
>How would the rust debugger be implemented?

We don't know yet, but the most straightforward way seems to be to reuse CLion infrastructure for LLDB/GDB.

**Comment by laktak** [comment]
I'm currently using Sublime with SublimeGDB. While it's nice it doesn't always work (crashes, doesn't keep breakpoints, etc.). So I have to fight the debugger as well...

I don't know CLion but whatever solution you come up with, please make sure it's really stable.

**Comment by sherry-ummen** [comment]
Sorry a newbie here. How do people debug rust code then? Or they don't ?

**Comment by Furyhunter** [comment]
You can use gdb or lldb directly. Recent versions of gdb support rust directly.

**Comment by sherry-ummen** [comment]
Thanks. Do you have an example on how to do it? And can it be used with IntelliJ rust?

**Comment by laktak** [comment]
@sherry-ummen if you are using sublime you can debug in your editor with SublimeGDB. It's not perfect but the best option at the moment.

**Comment by johnthagen** [comment]
Just FYI, [RustDT, an Eclipse based Rust IDE](https://github.com/RustDT/RustDT/blob/latest/documentation/Features.md#debugging-functionality) that appears to wrap GDB.  Might at least be a useful reference.

**Comment by sherry-ummen** [comment]
Meanwhile we can debug on Visual studio code https://sherryummen.in/2016/09/02/debugging-rust-on-windows-using-visual-studio-code/

**Comment by hgallagher1993** [comment]
For anyone thinking of using sherry-ummen's link, RustyCode appears to be abandoned there's been no updates in many months. [This](https://marketplace.visualstudio.com/items?itemName=kalitaalexey.vscode-rust) is a fork that fixes some bugs though. I haven't used the original in months and haven't even downloaded the fork so I don't how either works at the minute.

**Comment by chriskrycho** [comment]
I can confirm the fork works fairly well, and is (at least presently) fairly actively maintained. (I understand the limitations facing everyone involved here—I look forward to when IntelliJ Rust gets debugger support, but it's also a *ton* of work.)

**Comment by danielbodart** [comment]
I managed to get partial debug working using Remote GDB debug support in CLion and gdbserver. I can then see the stack traces but can't see the variables for what ever reason. Anyone got any suggestions?

**Comment by Nimamoh** [comment]
Would love some debug capabilities for intellij-rust.

**Comment by windoze** [comment]
I guess we can reuse CLion debugger module, or migrate this plugin to CLion.
I can see many plugins are available on both IDEA and CLion, so I guess these 2 products should have shared a large part of code base.

**Comment by matklad** [comment]
@windoze the plugin totally works with CLion, you don't need to migrate anything. 

I'm planing to start the work on the debugger soon :)

**Comment by csotiriou** [comment]
@matklad In CLion 2016.3.2 I installed the plugin, but the "New Project" dialog does not present the "Rust" option as it does in IntelliJ IDEA. Also, when trying to add a run configuration, the Rust option is also not there. Just mentioning it, because that makes rust-lldb config somewhat more difficult in CLion.

**Comment by matklad** [comment]
@csotiriou yep, the project management in CLion is not very convenient, because at the moment CLion does not support non-CMake project, and Rust obviously does not use CMake. The option is to use "import project from source" action.

**Comment by csotiriou** [comment]
@matklad Thank you for this answer.

I eagerly await for the debugger implementation. It's the only thing that is not straightforward to use, even when using the command line.

Thank you for this plugin.

**Comment by danielbodart** [comment]
Has anyone managed to get variables to show while debugging in CLion?

**Comment by matklad** [comment]
>Has anyone managed to get variables to show while debugging in CLion?

I did :) See the top comment for the current status: https://github.com/intellij-rust/intellij-rust/issues/535#issue-166047865

**Comment by johnthagen** [comment]
@matklad I'm curious how the final Rust debugger will interact, if at all, with the need for a CLion license. Will it be possible to debug Rust code on PyCharm Community, IntelliJ Community, or Android Studio using just the `intellij-rust` plugin?

I'd completely understand if JetBrains wanted to require that this part of CLion be licensed, since it pays good developers to work on CLion and needs to raise revenue.

**Comment by matklad** [comment]
@johnthagen this is something we have not figured out yet, both at the technical and the licensing side of things. 

It is certain that for the nearest future the debugger will be available in CLion, AppCode and maaaybe in RubyMine just because only these IDEs include `com.intellij.modules.cidr.debugger` module which we use. This is of course a shame, because it leaves out IDEA, PyCharm and other IDEs, and because Rust experience in CLion is not the best due to the interaction with CMake infrastructure. But of course this is just the beginning and we intend to fix these shortcomings :)

**Comment by lilianmoraru** [comment]
@hgallagher1993 The VS Code fork seems to be active, it even added support for RLS and works well(personally, I don't see much of a difference between the completion that `Intellij-Rust` and `VS Code + RLS` can give - for example, both can't help with variables passed through combinators like `map` and `and_then`).

As for `RustDT`, I can attest that the debugger works well(I tested it a while back).
Although, consider GDB 7.12 because it has better Rust support.

@matklad I installed the plugin but I cannot find the `Cargo debug` option, nor do I find the `crust` binary in the `Clion` directory.

**Comment by gilescope** [comment]
Seems like LLDB is being pushed by Apple over GDB. Hoping we can use CLion to support LLDB.

**Comment by matklad** [comment]
In the latest alpha, you don't need to install a separate debugger plugin or create a separate run configuration. 

Launching cargo command with debug "just works" (except that it doesn't handle a lot of stuff, like passing arguments to the debugged binary :) ). This required nightly cargo though.

https://zippy.gfycat.com/AcidicMediocreCoati.webm

**Comment by henryborchers** [comment]
No luck on my OS X or Windows machine. 

I first tried it on my Mac and the output message read, 
    Blocking waiting for file lock on build directory
    Finished debug [unoptimized + debuginfo] target(s) in 0.0 secs
I then got a popup stating "Debugging is not possible". 

When I tried it on my Windows computer, I got the same exact output messages but with no error popup message and my program didn't even execute. Is this only working on Linux for now?

**Comment by henryborchers** [comment]
D'oh. I forgot to set it to nightly Cargo. That's fixed but I'm now getting the message "NSUnbufferedIO is not set, output may be delayed" and I'm hitting no breakpoints.

**Comment by lilianmoraru** [comment]
The "Blocking waiting" problem seems to pop very often(Linux here).

**Comment by stephanbuys** [comment]
On MacOS Sierra I get, I've tried code-signing the bundled CLion gdb, but no luck yet.

```
macOS 10.12 has known issues running GDB due to System Integrity Protection.
Please switch the debugger to LLDB in Preferences.
```

**Comment by lilianmoraru** [comment]
I get "Debugging is not possible", "Can't find a binary to debug" with the latest nightly.

**Comment by matklad** [comment]
>I get "Debugging is not possible", "Can't find a binary to debug" with the latest nightly.

This can mean either of three things:

* you are using a stable cargo, which reports binaries only if they are not fresh (i.e. executing cargo clean before debugging should help)

* run configuration produced zero binaries (that's unlikely if you are debugging `cargo run` or `cargo test`). 

* run configuration produced more than one binary (for example, `cargo test`  excutes a ton of binaries at once). In this case, you need to pin the binary down by adding an explicit `--bin` or `--lib` argument. 

Sorry that this works awful at the moment: there still lots of stuff to do until the basic "it works"! :)

**Comment by lilianmoraru** [comment]
@matklad Sorry, it is a new machine and I forgot about the `nightly` thing(and you clearly mention it in the first post).
Thanks, setting rustup to nightly worked.

**Comment by henryborchers** [comment]
Has anybody got this to hit breakpoints on either on OS X or Windows?

**Comment by paul-tcell** [comment]
I created a project just like the debugger project in that webm above.   I'm getting that same **"Can't find a binary to debug"** error.

I'm using this nightly cargo build that I grabbed today:  `cargo-0.17.0-nightly (f9e5481 2017-03-03)`  (appears to be the latest `rustup install nightly` is pulling)

I'm using the master build of this plugin, downloaded today: `intellij-rust-0.1.0.1800`

The build produced one binary.

It's being run with `/Users/paul/.cargo/bin/cargo build --color=always --bin debuggertest`

**Comment by henryborchers** [comment]
I see the following message in the GDB console, "No source file named /Users/hborcher/rust/new_rust/src/main.rs." which is strange because that's the exact path of this source file.

Could this be related to why breakpoints are ignored when running this on my mac?

**Comment by minecrawler** [comment]
@henryborchers are you sure that path is correct? Shouldn't the users directory be lower case?

**Comment by henryborchers** [comment]
@minecrawler That's great news. I need to figure out what you are doing right and what I'm doing wrong. I'm using rustup nightly-x86_64-pc-windows-gnu, CLion 2017.1.1. Mingw64 (msys2) which uses GDB 7.12.1 on Windows 7

**Comment by dragostis** [comment]
Sorry for popping in like this, but how does Android Studio's native debugger work? Wouldn't it be possible to do the same with IntelliJ?

**Comment by gilescope** [comment]
Just switched to a 'fresh' OSX install and lldb debugging is now working well with CLion. :-)

But like all good things, I want / need more. Specifically:

Really need String / &str to be viewable in the debugger as their contents rather than implementation. Same for vecs and hash maps. In Visual Studio you could create debugging visualisers and I can see it's possible to do so in llvm. In fact it looks like rust already has some nice formatters that can be applied in rust/src/etc/lldb_rust_formatters.py  - I've just not quite figured out how to plumb this all together. Has anyone already managed this yet?

**Comment by gilescope** [comment]
Still tinkering but this for example seems to nicely format Strings:
type summary add --summary-string "${var.vec.buf.ptr.pointer.__0}" alloc::string::String 

(from the excellent docs at https://lldb.llvm.org/varformats.html )

**Comment by lilianmoraru** [comment]
@gilescope You should have the `rust-gdb` and `rust-lldb` binaries in your toolchain.
These binaries load `gdb`/`lldb` with the appropriate formatters.

**Comment by henryborchers** [comment]
@lilianmoraru , I figured rust-gdb and rust-lldb would be installed by default by rustup. How do I check that I have these binaries?

**Comment by henryborchers** [comment]
Oh snap! I just tried it agin. it's actually working now. Awesome!

**Comment by gilescope** [comment]
While things didn't work out of the box for me, by adding a ~/.lldbinit file with the following did the trick:
```
comm scri imp -r ~/your-path-here/etc/lldb_rust_formatters.py
type summary add --no-value --python-function lldb_rust_formatters.print_val -x \".*\" --category Rust
type category enable Rust
```
 
(I also had to chmod the lldb_rust_formatters.py file but this could just be that I was picking my copy up from a clone of rust source.)

You can confirm it works by typing "type category enable Rust" into the lldb prompt - if it doesn't error that's a good sign. :-)
Hope this helps someone.

**Comment by stephanbuys** [comment]
The note above works well on CLion/MacOS.
I still get regularly get `Blocking waiting for file lock on build directory` errors though. 
Looking at the output of `lsof  ./target/debug/.cargo-lock` I can see that occasionally cargo doesn't exit even after the process has excited, not sure why. 

Currently killing the cargo processes and manually deleting the lock file allows me to rerun the debug command.

**Comment by farodin91** [comment]
@matklad Any progress in supporting debug in Intellij IDEA?

**Comment by matklad** [comment]
@farodin91 sorry, not a lot of progress on debugger front recently :(

**Comment by pabloa** [comment]
Please re evaluate the importance of adding a realistic debugging solution to Rust.

The "big masses" of developers interested in Rust are probably not comfortable/interested on C/C++. Perhaps we are more interested in using Rust on projects that can be done with java, go, Scala or similar. 

All these CLion cluttering asking me about C++ binaries, make/cmake drama, refusing to recognize my Rust project or even opening is sad. I cannot use IntellIJ because it refuses to run my Rust program in debug mode.

We NEED a tool to easy debugging Rust code or it is simply not possible to run any serious application in Rust without a C/C++ background. 

If you guys are truly interested in wide Rust adoption, please work in the debug front. It is more important than adding more features to the library or even fixing bugs in std lib.

I spent 4 weeks learning Rust and I was prepared for the tough learning curve ahead me. I wrote a program to send/receive messages and I am just tired to debug it using print!() to the console. Clearly I cannot use these to write bigger apps. Any more serious application is not possible and I will not learn C++ tools stuff to disabled them so I can use a debugger :)

All these saying with nothing but Rust love. Still interested in Rust though! Not ready for a medium project written for not genius, C junior, short time developers.

**Comment by Redrield** [comment]
> The "big masses" of developers interested in Rust are probably not comfortable/interested on C/C++

That seems to go against all that Rust is trying to do. Rust isn't trying to be another application level language like those that you've listed. It's a system language, and it has been since it was created. 

If you don't like fiddling with CLion to get embedded debugging to work, realistically, you can just open a bash/wsl shell session, `cargo build` and use `rust-gdb` or `rust-lldb` to debug the program. Using gdb, at least, from the command line is a lot easier than it might seem. 

Not to mention the fact that if you truly _need_ an embedded debugger, you could also simply set up an environment with Visual Studio Code, and use its built in LLDB debugger.

**Comment by pabloa** [comment]
Thank you @Redrield for your answer. I will try the command line approach you suggested.

I want to use Rust to write a server app and experiment how this program would perform if it is written in a system language without GC. It seems high level enough to have a similar experience than writting java code, except the debugging part that I feel it is too complicated.

**Comment by sssilver** [comment]
@Redrield I don't think you fathom the damage this kind of sentiments do to the dissemination of the language. I'm not interested in debating opinions here. Luckily, we have the [2017 Rust Roadmap](https://github.com/aturon/rfcs/blob/roadmap-2017/text/0000-roadmap-2017.md) to guide us:

> 1. Rust should have a lower learning curve
> 2. Rust should have a pleasant edit-compile-debug cycle
> 3. Rust should provide a solid, but basic IDE experience
> ...

These are no more no less the first 3 priorities that the core team has established as focus areas for 2017 in order for Rust to succeed and get to the next level.

The worst thing we can do for a language as new and, dare I say, as unusual as Rust, is labeling it as niche, and tagging it with spooky phrases like "systems language". The very first paragraph of the [official Rust book](https://doc.rust-lang.org/book/second-edition/) (current edition) specifies:

> Welcome to “The Rust Programming Language,” an introductory book about Rust. Rust is a programming language that’s focused on safety, speed, and concurrency. Its design lets you create programs that have the performance and control of a low-level language, but with the powerful abstractions of a high-level language. These properties make Rust suitable for programmers who have experience in languages like C and are looking for a safer alternative, _as well as those from languages like Python who are looking for ways to write code that performs better without sacrificing expressiveness_.

I hope my comment helps @pabloa stay inspired and on point -- the way you use Rust is exactly one of the the right ways to use it, and your expectation of having a normal debugger is something many (if not most) of us share. Current debugging is abysmal.

Can't wait to have a proper debugger for a language I use to write curses-based Tetris clones!

**Comment by pabloa** [comment]
Idea is supporting officially Rust :D
https://blog.jetbrains.com/blog/2017/08/04/official-support-for-open-source-rust-plugin-for-intellij-idea-clion-and-other-jetbrains-ides/

Please comment asking for a working debugging experience :+1:

**Comment by gilescope** [comment]
I think a lot more work needs to be done lower down the stack than at the intelli-rust level to get something as slick as us non-C++ devs are used to. A great debugging experiance is a joy to behold, and maybe there's an expectations gap - the C++ guys haven't really had as great a debugging experiance as the manage languages have. We come expecting a lot. It will come, but it takes time. For me personally not coming from a C++ background means there's a huge learning step to figure out when and which shared pointer should be used. I've taken a lot longer than 4 weeks! But while I've been learning the language it's been great to see just how much leaps and bounds the IDE, language and ecosystem have come on. But to set expections, debugging on CLion or through rust_lldb is sketchy, for now println("{:#?}") is better in some ways as you can invoke the Debug trait rather than relying on the visualisers. Lots of work to do, but mostly I think it's not intelli-rust, but rust-core/llvm and lldb where it needs to be done. 
(Again many thanks for your blood, sweat and tears making intelliRust - it's getting more awsome by the week!)

**Comment by pabloa** [comment]
I finally found a way to make Rust plugin and the debugger work with CLion. 

The steps are:

- Make sure your Rust project works with Cargo. Try to compile it and call the executable. Verify your program starts.
- Download CLion. And install it somewhere. Do not insist (as today Aug 7th, 2017) with any other Jetbrains IDE. Debug works only in CLion. I tried with the paid version of the CLion IDE. 
- Launch it. Configure LLDB as your debugger (Please report if it works when GDB is set).
- Install Rust and Toml plugins. Restart IDE.
- Create a fake **CMakeLists.txt** file on the root of your Rust project with this content:
```cmake
cmake_minimum_required(VERSION 3.8)
project(untitled)

set(CMAKE_CXX_STANDARD 11)
```
This cmake file does not make sense. It does not matter. Carry on.
- In CLion, go to Open Project, and open your project by clicking in the **CMakeLists.txt** file.
- You could get an error message saying something like:
```
Unknown Module Type
Cannot determine module type ("JAVA_MODULE") for the following module:"yourProject"
The module will be treated as a Unknown module.
```
I ignored the error. 
- open your main.rs file, look for the first executable line after the ``` fn main() {``` one. Set a breakpoint as usual by clicking on the left of the line.
- Admire that beautiful breakpoint so missed, so dreamed...
- Right click in the play symbol at the left of the main() line. Choose the option Debug.
- After 1 or 2 seconds: Behold! You application is launched with GDB and stopped in your breakpoint!
- Celebrate with beer or coffee.
- hack, debug, deliver, collect $$$

**Comment by vmaks** [comment]
I am looking for debugger support!

**Comment by lilianmoraru** [comment]
@pabloa I use the debugger without any `CMakeLists.txt` files(in my own projects).
I know you can't open your first Rust project without a `CMakeLists.txt` file - I open another C++ project and after that you can open any Rust project through `File -> Open`.

**Comment by vmaks** [comment]
@pabloa Thank you very much! It really works!

**Comment by gilescope** [comment]
If there was one thing that would make Rust debugging more effective, it would be having the debugger pause on panics / unwinds. They _shouldn't happen_ in normal control flow, so it would be great if we could hint to lldb that there should always be a breakpoint when an unwind starts.

**Comment by kellerkindt** [comment]
Having no debugger in IntelliJ IDEA is quite a shame. I am sitting here on a multi module project, that cannot be opened with CLion because they are multiple modules (and nobody wants CMakeFiles in a rust module), but cannot be debugged with IntelliJ IDEA **Ultimate** because... well.. nobody knows.

I really like IntelliJ and i'ts a pleasure to work with, but this is just sad.
Looking forward to be able to debug in IntelliJ... eventually.

**Comment by matklad** [comment]
>cannot be opened with CLion because they are multiple modules 

Hey, starting with https://github.com/intellij-rust/intellij-rust/pull/1791 (Monday's release) you can open may projects in CLion at once :) 

> cannot be debugged with IntelliJ IDEA Ultimate because... well.. nobody knows.

Ultimate does not have the necessary debugging infrastructure unfortunately (like, CLion bundles a specific LLDB).

**Comment by stephanbuys** [comment]
Similar concern this side, when doing frontent/rocket development I would like to use Webstorm and still have all the rust capabilities available... 
@matklad Would it be possible to do proper web debugging in CLion with the capabilities you mentioned above?

**Comment by Cyberax** [comment]
I'm getting an error popup:
```
Error running <project>
com/jetbrains/cidr/cpp/CPPToolchains
```
Interestingly enough, I don't see any exceptions in the IDEA log.

**Comment by matklad** [comment]
@Cyberax see https://github.com/intellij-rust/intellij-rust/issues/1923

**Comment by avkonst** [comment]
Is current suggestion to use CLion, which is paid product, long term solution? or the Rust plugin will get native support?

**Comment by minecrawler** [comment]
@avkonst afaik, the plugin is compatible with all JetBrains products. However, the debugger only seems to work with selected products, for example CLion, which already provides the necessary infrastructure.

**Comment by avkonst** [comment]
@minecrawler Does it mean that the full benefit of the plugin can be leveraged only with CLion ? Are there plans to have the same for Intellij Idea?

**Comment by minecrawler** [comment]
@avkonst Maybe @matklad can help

**Comment by lilianmoraru** [comment]
@avkonst The plugin is fully usable, it's just that the CLion debugger is not available with IntelliJ.

**Comment by makslevental** [comment]
@matklad 

not sure what i'm doing wrong but i'm getting this error in CLion when trying to debug

![image](https://user-images.githubusercontent.com/5657668/33268344-ff8bdf42-d34a-11e7-9c87-f54005de8249.png)

and yet

![image](https://user-images.githubusercontent.com/5657668/33268565-b39adf60-d34b-11e7-90f4-a9d052ba8248.png)


here is my folder structure


![image](https://user-images.githubusercontent.com/5657668/33268328-f015cf32-d34a-11e7-8e9e-f421fe23f028.png)

**Comment by AurevoirXavier** [comment]
I have the same problem with @makslevental .
`cargo --version
cargo 0.24.0-nightly (5bb478a51 2017-11-29)`
Clion: 2017.2.3
System: High Sierra
Although there is only a `main.rs` file in `src/`, I specify `run --package plasma --bin plasma` .
But it still show me this error.

**Comment by UYEONG** [comment]
I have the same problem too.

![2017-12-06 11 12 29](https://user-images.githubusercontent.com/1321707/33640945-d338f7ee-da76-11e7-8491-9b5e7104150d.png)

---

![2017-12-06 11 13 28](https://user-images.githubusercontent.com/1321707/33640961-e1c5cca6-da76-11e7-9360-bf9a2b6d106c.png)

---

![2017-12-06 11 13 40](https://user-images.githubusercontent.com/1321707/33640969-e6c6ad1a-da76-11e7-994f-1602d9699338.png)

**Comment by RobinRheem** [comment]
@UYEONG Why not help the project! We're counting on you 👍

**Comment by chrisballinger** [comment]
@AurevoirXavier @UYEONG  I have this same error message "Debugging is not possible. More than one binary produced". Here's a sample project: https://github.com/chrisballinger/sds011-rs

It seemed to work okay for a simple binary project but once I remade it as a lib+binary it stopped working. I can still use the CLion Run command, just not Debug.

* cargo 0.23.0
* clion 2017.2
* rust plugin: 0.2.0.2081

**Comment by AurevoirXavier** [comment]
@chrisballinger Hello, what your *simple binary project* mean? Even just `cargo new xxx --bin`, I still get that error...

**Comment by chrisballinger** [comment]
@AurevoirXavier It was a `cargo new xxx --bin` at first and it worked, then I did `cargo new xxx` and moved the source into there and manually created a new main file in `src/bin/sds011.rs`. Perhaps there was a regression in one of the last two updates of the rust plugin?

**Comment by Pzixel** [comment]
So it's not possible to debug Rust project in interlij community edition? Because the only way to make it work is pay for CLion? Damn, it's quite not good for non-Java developers who's trying to learn some Rust, but it only has VSCode plugin that eventually stops to work and IntelliJ plugin that doesn't support debugging...

**Comment by AurevoirXavier** [comment]
@makslevental  @UYEONG  @chrisballinger Try Clion 3.1 EAP and set toolchains to LLDB. It work for me now!

**Comment by makslevental** [comment]
@AurevoirXavier thanks will do

**Comment by UYEONG** [comment]
@AurevoirXavier t's working for me too! thanks.

**Comment by U007D** [comment]
I'm getting this same error as well (I am on Ubuntu):
![](https://user-images.githubusercontent.com/1321707/33640969-e6c6ad1a-da76-11e7-994f-1602d9699338.png)

My project is not the simplest (multiple sub-crates using workspaces) with the main crate being lib + bin, but seeing as so many others are having this same issue, I'm hoping the project's layout is not the issue (at least it's not the only issue! :)).

* I'm invoking a `cargo test --lib --all --tests` configuration by clicking the Debug button.
* Toolchains | Debugger is set to 'Bundled LLDB` `✓Version 5.0.0`

CLion:
```
CLion 2017.3.1 Build #CL-173.4127.32, built on December 26, 2017
JRE: 1.8.0_152-release-1024-b8 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.13.0-19-generic
```
Intellij-Rust:
```
Version: 0.2.0.2083
```

`cargo`: `cargo 0.25.0-nightly (930f9d949 2017-12-05)`
`rustc`: `rustc 1.24.0-nightly (5165ee9e2 2017-12-22)`

I'm assuming I don't need to downgrade to CLion 3.1 EAP--please correct me if I'm wrong.  Any ideas?

**Comment by norru** [comment]
@pabloa's approach above has been working for me for many months now, with Linux/gcc toolchain at least. Variable inspections are very rough on the edges though.

CLion + Rust plugin is getting there but a LONG way from mainstream-ready.

**Comment by perlindgren** [comment]
I'm experimenting with debugging Rust on embedded (Nucleo STM32F401re) using openocd and arm-none-eabi-gdb. I have it somewhat working. Here is the current status.

1. The program is not loaded. (I had to use the pause and in the gdb console load it, the `file` has the right path, so it loads correctly)
2. Breakpoints do not work, setting them in the IDE has no effect, stepping works and even (surprisingly) run to line. Setting breakpoints in the gdb console works fine.
3. Variables are sometimes not shown. When they work, they are also updated in the source file. However mouse overs never work.
4. Whatching
![untitled -clionprojects-rtfm-app - -examples-bare0 rs rtfm-app - clion_009](https://user-images.githubusercontent.com/4089704/35020529-df710a52-fb2c-11e7-94d7-1f6e81ac7700.png)
 local (stack) as well as static (heap) variables always work.

I'm on the latest nightly xargo/cargo and nightly rustup channel. Arch linux with the CLion December 26 build. The bebug config:
/usr/bin/arm-none-eabi-gdb 
(GNU gdb (GDB) 8.0.1)
'target remote' args: :3333
Symbol file:/home/pln/CLionProjects/rtfm-app/target/thumbv7em-none-eabihf/debug/examples/bare0
Sysroot: 
(Empty)

To me it seems not so far from working and the Rust support in CLion looks promising (perhaps the best amongst RustDT (not supported), and vscode (RLS based)). So a couple of questions/remarks.
.gbinit will run before the 'target remote' from the plugin is launched. I tried to use that to load the binary, but I did not get that to work. It would be better with a 'classic' script option, allowing you to override the default connection procedure. Regarding the breakpoint integration, I have no clue what is wrong, as there seems to be some connection in between the IDE and gdb. (Rust uses C++ style namespaces/symbols, so that should not be a problem.) Another observation, if you try run to line, on a line not reachable, the debugger will not halt even if you press pause. I also tried using the OpenOCD Dowload & Run plugin, but it did not do the trick as being tailored to C/C++ (It could be a good stard for a dedicated Rust embedded debug plugin though...). 

Any feedback appreciated. 
/ Per
(The example project I based this experiment on is found at https://gitlab.henriktjader.com/pln/rtfm-app)
![image](https://user-images.githubusercontent.com/4089704/35020551-f9e3b9f2-fb2c-11e7-8354-ee0d0e1159d0.png)

**Comment by tdbgamer** [comment]
Is debugging support even planned for IntelliJ IDEA? If everyone has to buy CLion to use Rust, they're just going to switch to VSCode or other alternatives.

**Comment by gilescope** [comment]
Well they have to make money some how. I'd be happy to pay for a debugger
that works well :-)

**Comment by tdbgamer** [comment]
@gilescope 
Don't get me wrong, I'm not whining or expecting them to do anything for me for free. But...

They don't make money from developers buying their IDEs. They make money when developers _use_ their IDE and force their company to buy licenses. That's why there are so many community editions (also so they can let contributers maintain their code for them).

I would never pay $ 199.00/year for just debugging ($15-$20 plugin? I'd absolutely do it). Instead my team and I will use different editors and have no incentive to ask our bosses to buy IntelliJ licenses... Just sayin. Same reason Microsoft gives everything away for free to college students. It's a pretty good business model.

**Comment by Pzixel** [comment]
@tdbgamer I think plugin guys have never to deal with pricing 😄 And it seems that IntelliJ just doesn't provide infrastructure required to make debugger work. It's sad, but it's a mistake of IDE guys, not plugin ones.

**Comment by tdbgamer** [comment]
@Pzixel
Sure I understand that. But I didn't know if they planned on implementing a workaround in the plug-in or not. Probably the answer is no based on the responses. Hopefully the IDE people will see this then.

**Comment by sssilver** [comment]
Can we escalate this and get some sort of an official response?

**Comment by farodin91** [comment]
@Undin Is there any progress on the debugger side?

**Comment by Undin** [comment]
> Is debugging support even planned for IntelliJ IDEA?

Currently, the debugger is supported only in CLion because it has all the necessary infrastructure for working with LLDB, GDB and native code. We don't have immediate plans to move it to IDEA

**Comment by sanmai-NL** [comment]
@Undin: I think you response is clear and your statement is nothing new really, thanks. But could you or @matklad perhaps change the issue title to add ‘... in CLion’ to prevent any confusion?

**Comment by Pzixel** [comment]
@sanmai-NL Clion already have implemented debugger? What do you mean then?

**Comment by sanmai-NL** [comment]
Apparently that implementation isn’t completely done yet. https://github.com/intellij-rust/intellij-rust/issues/535#issuecomment-347179169 . If @matklad: now thinks otherwise, the whole issue can be closed instead.

**Comment by vlad20012** [comment]
@sanmai-NL Note that matklad leaved JetBrains and now don't work on IntelliJ-Rust

**Comment by sanmai-NL** [comment]
@vlad20012: okay, mentioned him since he is the issue reporter. 🙂

**Comment by gilescope** [comment]
For anyone interested in improving rust debugging, there is work going on here: https://github.com/tromey/lldb/tree/rust

Hopefully this will lead to IDEs being able to provide a richer debugging experience once it is complete.

**Comment by gilescope** [comment]
Just wanted to say how much faster the debugger is now compared to a year ago. Really impressed. Just need String to appear as a &str and I will be at peace 👍

**Comment by Undin** [comment]
> Just wanted to say how much faster the debugger is now compared to a year ago.

Thanks to CLion team!

> Just need String to appear as a &str and I will be at peace

It's already implemented (see #2494) and will be landed in next release

**Comment by gilescope** [comment]
December so soon? :-) big thanks!

**Comment by andrewdavidmackenzie** [comment]
The link in description (https://teamcity.jetbrains.com/repository/download/IntellijIdeaPlugins_Rust_BuildDebugger/.lastSuccessful/intellij-rust-debugger-0.0.1.%7Bbuild.number%7D.zip) is broken.

Searching TeamCity for "rust debug", "rust debugger", "intellij-rust-debugger" doesn't show anything.

Where can I download the plugin to add debugging capabilities to rust, either for IntelliJ with rust plugin (preferred) or CLion?

thanks

**Comment by U007D** [comment]
The latest CLion + [IntelliJ Rust](https://intellij-rust.github.io) will give you a debuggable Rust development environment.  Note: CLion specifically is required.

**Comment by Undin** [comment]
> The link in description (https://teamcity.jetbrains.com/repository/download/IntellijIdeaPlugins_Rust_BuildDebugger/.lastSuccessful/intellij-rust-debugger-0.0.1.%7Bbuild.number%7D.zip) is broken.
> Searching TeamCity for "rust debug", "rust debugger", "intellij-rust-debugger" doesn't show anything.

@andrewdavidmackenzie yeah, it's the outdated link in "Old content" section.
Now you don't need any other plugin except "intellij-rust".

> Where can I download the plugin to add debugging capabilities to rust, either for IntelliJ with rust plugin (preferred) or CLion?

You can install it in `Plugins` settings in IDE or find in the [plugin repository](https://plugins.jetbrains.com/plugin/8182-rust). As before, debugging is available only in CLion

**Comment by andrewdavidmackenzie** [comment]
I got it working after adding localhost entry to /etc/hosts

The description made me think it might work in IntelliJ, my mistake. Thanks!

**Comment by johnthagen** [comment]
@perlindgren Were you ever able to make progress on CLion + STM32/OpenOCD? I am working trying to also get that to work (opened a [Rust Users thread about it](https://users.rust-lang.org/t/embedded-rust-how-to-debug-stm32-openocd-from-clion-intellij-rust/19062)).

**Comment by crusty-dave** [comment]
**Pricing reality check**

IntelliJ IDEA actually cost more than CLion.  So if I can only get debugging in CLion then they (JetBrains) lose money because I would drop my IDEA subscription.

I didn't realize that IDEA didn't support a debugger when I bought it - my bad.  I already have Visual Studio so I wouldn't have thought to purchase CLion over IDEA.

The good news is so far I haven't needed a debugger for Rust, but if / when I do, I will have to make a decision if I don't want to use a command line debugger.

**Comment by avkonst** [comment]
As an other option: Visual Studio Code native code debugger works just fine
for rare cases, when I need to debug Rust. It has got watches, breakpoints,
interactive variables states inspection, etc... So, I use IDEA with rust
plugin and debug in VS Code rarely.

**Comment by vlad20012** [comment]
@crusty-dave
> IntelliJ IDEA actually costs more than CLion. So if I can only get debugging in CLion then they (JetBrains) lose money because I would drop my IDEA subscription.

It's hard to make debugger work in IDEA due to technical reasons. Ideally, our plugin should be fully functional in community IDEA (for free). So it's not a business/marketing issue.

> I didn't realize that IDEA didn't support a debugger when I bought it - my bad. I already have Visual Studio so I wouldn't have thought to purchase CLion over IDEA.

If you accidentally bought the wrong product, I think our support can solve this issue.

**Comment by johnthagen** [comment]
Also, if you work on a published, open source project, you can apply for a free CLion license.

https://www.jetbrains.com/buy/opensource/

**Comment by zhuxiujia** [comment]
i use  rustup install stable-gnu and Preferences > Build, Execution, Deployment > Toolchains set MinGW.
but  start debug  it tell me  "MSVC toolchain is not supported for debugging,Please use GNU toolchain"!

--- just same you say do it, It doesn't work at all !!!!!!!!!!!!

**Comment by zhuxiujia** [comment]
vscode support use LLDB to debug  , 
 Way intellij-rust  not support  LLDB!

**Comment by Marlinc** [comment]
I would love to see debugging support in IntelliJ IDEA

**Comment by Pzixel** [comment]
I'm working in CLion for a long time now and I discussed this matter a lot. In a nutshell: there won't be any debugger in IDEA in foreseeable future because IDEA doesn't have required toolchains/internal APIs/so on. And I think if will appear in some kind of Rust-only IDEA-based IDE (just like pycharm/goland/...) rather than in IDEA itself. So there is kinda nothing to discuss: it won't change because of someone writings. Just skip by and take CLion/VSCode, whatever you can, and wait for rust IDE.

Since there will be `rust-analyzer` in a near future, IDEA may be replaced relatively painless.

**Comment by crusty-dave** [comment]
OK, I really need to debug some Rust code, so I upgraded to all packages so that I can build / debug using CLion.  Only to find out that it requires the gnu version of Rust, why?  It turns out the gnu version won't build ring, I have already had my share of issues getting it to build with MSVC, so I am not happy right now.  

```
The system cannot find the file specified. (os error 2)', ~\.cargo\registry\src\github.com-1ecc6299db9ec823\ring-0.13.5\build.rs:642:9
```

**Comment by crusty-dave** [comment]
To put it another way, we are using clang with MSVC, we *will not* ship with gnu - nor should we be expected to debug with gnu.  CLion debugging really needs to work with clang.  Hopefully there is some way to get this to work???

**Comment by andrewdavidmackenzie** [comment]
I use CLion with standard rust and debug just fine, and debugger support
had got gradually better and better (variables/struct display, etc).

**Comment by crusty-dave** [comment]
> I use CLion with standard rust and debug just fine, and debugger support had got gradually better and better (variables/struct display, etc).

On Windows?

I get "MSVC toolchain is not supported.  Please use GNU toolchain.", so it doesn't work for me.  As I stated above, if I try to build with the GNU toolchain, ring fails to build.  I don't even want to begin to look into that turd as it was difficult enough to get ring to build correctly with MSVC.

**Comment by andrewdavidmackenzie** [comment]
No. MacOS :-(

**Comment by crusty-dave** [comment]
Well, maybe that is an option for me, I do have a Mac available.  Thanks for that insight!

**Comment by Undin** [comment]
> I get "MSVC toolchain is not supported. Please use GNU toolchain.", so it doesn't work for me.

@crusty-dave Yeah, it's expected right now.
CLion didn't support debugging of binaries built with msvc toolchain until recently. It released **experimental** debugger for msvc binaries with the latest EAP. Unfortunately, there are some technical issues to use it with Rust msvc binaries at this moment but we are working on it

**Comment by ortem** [comment]
@crusty-dave Rust debugging in CLion should work pretty well on Windows with GNU toolchain. All you have to do is:
1. `rustup toolchain install stable-x86_64-pc-windows-gnu`
2. `rustup default stable-x86_64-pc-windows-gnu`
3. Install Cygwin x86_64 or MinGW-W64 with supported gdb version (7.8.x-8.2.x)
4. Select your Cygwin/MinGW toolchain in `File | Settings | Build, Execution, Deployment | Toolchains`

**Comment by axos88** [comment]
What about Intellij IDEA? Can I get it to work there?

**Comment by Undin** [comment]
@axos88 No. Debugger is available only in CLion at this moment

**Comment by ShadowRZ** [comment]
July 28, 2019 3:39 PM, Arseniy Pendryak wrote:

> No. Debugger is available only in CLion at this moment

Hmmmm......
Since debugging Rust apps requires CLion Debugger, can I think that this is a CLion-only extension?

**Comment by Undin** [comment]
@ShadowRZ No
Only debugger, profiler and Valgrind memcheck features are CLion-only.
All other features are available in all IDEs

**Comment by axos88** [comment]
So the recommendation is to use CLion instead of IntelliJ if I want to debug a rust app?

**Comment by axos88** [comment]
Or is this the recommendation in general, if one has access to the whole product pack?

**Comment by ortem** [comment]
@axos88 Unfortunately, Rust debugging doesn't work in IntelliJ IDEA, because for now native languages (C/C++/Rust) debuggers only available in CLion, not in IDEA. So you **have to** use CLion if you need Rust debugging support.

**Comment by GrumpyMetalGuy** [comment]
Hi, is there any plan to add support for an option to not step into the core Rust scripts when debugging?

**Comment by ortem** [comment]
@GrumpyMetalGuy Hi, could you please describe your request in detail and provide an example?

**Comment by norru** [comment]
FIXED or WONTFIX?

**Comment by Undin** [comment]
@norru debugging is supported in CLion.
If you are looking for it in other IDEs, see #3453

**Comment by norru** [comment]
@Undin This is fine, I just wanted to have an idea of what the plans were. "Debugger will be only available in CLion for the foreseeable future" is an acceptable answer.

**Comment by bugproof** [comment]
what about this https://www.forrestthewoods.com/blog/how-to-debug-rust-with-visual-studio-code/ this workaround seems to be using msvc debugger

**Comment by crusty-dave** [comment]
I am able to use the Visual Studio debugger with Rust on Windows 10 with no issues.  If the process is running, just attach to it and set break points.  I haven't tried actually starting the Rust process in Visual Studio, I would imaging that just takes a bit more setup.

Currently I find CLion *only* useful as an IDE.  The Rust plug-in doesn't support changing the cmd prompt for running the compiler.  In my case I have a build script that needs the resource compiler, so I need it to cargo build under a Visual Studio x64 Native Tools Command Prompt, which I cannot currently setup with the Rust plugin (at least the last time that I tried).

**Comment by serid** [comment]
Not sure if this is a bug, but debugging Rust code using **stable-x86_64-pc-windows-gnu Rust toolchain** does not work when MinGW/Cygwin is not the default **CLion toolchain**.
![image](https://user-images.githubusercontent.com/30107608/111728596-3caa9080-887e-11eb-97e6-810325c333dd.png)

Swapping around the **CLion toolchain** fixes the issue. This info might be of use for future readers.
![image](https://user-images.githubusercontent.com/30107608/111728790-a7f46280-887e-11eb-8df2-231a6c6b51ec.png)

**Comment by Undin** [comment]
@serid it's not a bug, it works as expected. `Toolchains` settings are implemented as a way to reuse different sets of toolchain settings across different projects but only default one is used for a particular project
